dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -485,7 +485,17 @@ object DistriOptimizer {
       parameter.narrow(1, start + 1, length).copy(weights(pid))
     })
 
-    trainedModel
+    if (null != result) {
+      // Copy trainedModel to result
+      val resultParameter = result.getParameters()._1
+      require(resultParameter.nElement() == parameter.nElement(),
+        "result's parameter should have the same size with train model's parameter")
+      resultParameter.copy(parameter)
+      nn.Utils.recursiveCopyState(trainedModel, result)
```

Comment:
I think you can define a recursiveCopyModule method, which in turn takes care of both the parameter setting and state copy.